src/Microsoft.Health.Fhir.Liquid.Converter/Processors/BaseProcessor.cs

```diff
@@ -12,28 +12,52 @@
 using Microsoft.Health.Fhir.Liquid.Converter.Exceptions;
 using Microsoft.Health.Fhir.Liquid.Converter.Models;
 using Microsoft.Health.Fhir.Liquid.Converter.OutputProcessors;
-using Newtonsoft.Json;
+using Microsoft.Health.Fhir.Liquid.Converter.Telemetry;
+using Microsoft.Health.Logging.Telemetry;
+using Newtonsoft.Json.Linq;
 
 namespace Microsoft.Health.Fhir.Liquid.Converter.Processors
 {
     public abstract class BaseProcessor : IFhirConverter
     {
-        protected BaseProcessor(ProcessorSettings processorSettings)
+        private readonly ITelemetryLogger _telemetryLogger;
+
+        protected BaseProcessor(ProcessorSettings processorSettings, ITelemetryLogger telemetryLogger)
         {
             Settings = EnsureArg.IsNotNull(processorSettings, nameof(processorSettings));
+            _telemetryLogger = EnsureArg.IsNotNull(telemetryLogger, nameof(telemetryLogger));
         }
 
         public ProcessorSettings Settings { get; }
 
+        protected ITelemetryLogger TelemetryLogger { get => _telemetryLogger; }
+
         protected virtual string DataKey { get; set; } = "msg";
 
         public string Convert(string data, string rootTemplate, ITemplateProvider templateProvider, CancellationToken cancellationToken, TraceInfo traceInfo = null)
         {
             cancellationToken.ThrowIfCancellationRequested();
-            return Convert(data, rootTemplate, templateProvider, traceInfo);
+            string result;
+            using (ITimed totalConversionTime = TelemetryLogger.TrackDuration(ConverterMetrics.TotalDuration))
+            {
+                result = InternalConvert(data, rootTemplate, templateProvider, traceInfo);
+            }
+
+            return result;
+        }
+
+        public string Convert(string data, string rootTemplate, ITemplateProvider templateProvider, TraceInfo traceInfo = null)
+        {
+            string result;
+            using (ITimed totalConversionTime = TelemetryLogger.TrackDuration(ConverterMetrics.TotalDuration))
```

Comment:
What is the best way to associate these metrics to the operation status, i.e., pass/fail. Should we consider adding dimensions to also capture whether the overall conversion succeeded or failed.
If say the conversion fails early (in case one of the inner steps throws early), the duration could be short and give incorrect impressions on the performance.

Reply:
I agree with your suggestion. So far, this change is to add basic touch points to evaluate the duration of transformation and its each steps. I also believe that we can extend this to have more dimensions and create metrics on our backend to monitor. A precondition of doing that would be a general overview of metrics/alarms strategy. We also need to consider the implication and recommendations for external customers when we update the OSS. I would think another task to wrap up that scope. :)